### PR TITLE
Usb004 secure login

### PR DESCRIPTION
**Pull request by rdgrv**
Password gets encoded and decoded
SQL script uploads an encoded version of the default password (12345678) to the default users
UserService loginUser unit tests

### CODE REVIEW

File: LearningTool/src/main/java/main/services/UserServiceImpl.java [hunk 1]
```diff
@@ -33,22 +34,38 @@ public User findById(Long id) {
...
+        user.setPassword(encodePassword(user.getPassword()));
+        return userRepository.save(user);
+    }
+
+    @Override
+    public User loginUser(UserDTO userDTO) {
+        LOGGER.info("Logging a user in...");
+        User user = userRepository.findByEmail(userDTO.getEmail()).orElseThrow(NullPointerException::new);
+        if (decodePassword(userDTO.getPassword(), user.getPassword())) {
+            return user;
```
Comment (AnnaDeak) [inline LearningTool/src/main/java/main/services/UserServiceImpl.java]: Logging

File: LearningTool/src/main/java/main/services/UserServiceImpl.java [hunk 2]
```diff
@@ -33,22 +34,38 @@ public User findById(Long id) {
...
+        } else {
+            throw new NullPointerException();
```
Comment (AnnaDeak) [inline LearningTool/src/main/java/main/services/UserServiceImpl.java]: logging

File: LearningTool/src/test/UserServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+        userService.loginUser(userDTO);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void loginUser_noUserFound(){
+        String password = "testPassword#123";
+
+        UserDTO userDTO = new UserDTO();
+        userDTO.setPassword(password);
```
Comment (AnnaDeak) [inline LearningTool/src/test/UserServiceTest.java]: remove

File: LearningTool/src/test/UserServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Before
+    public void init() {
+        MockitoAnnotations.initMocks(this);
+        userService = new UserServiceImpl(userRepository, roleRepository);
+    }
+
+    @Test
+    public void loginUser_successful(){
+        String password = "testPassword#123";
```
Comment (AnnaDeak) [inline LearningTool/src/test/UserServiceTest.java]: private static

File: LearningTool/src/test/UserServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import org.mockito.junit.MockitoJUnitRunner;
+import org.springframework.security.crypto.bcrypt.BCrypt;
+
+import java.util.Optional;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class UserServiceTest {
```
Comment (AnnaDeak) [inline LearningTool/src/test/UserServiceTest.java]: same folder structure

**Comment by anikofenyvesi** [comment]
Hello, just some comments to the unit test part:
1, According to the maven directory structure (https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html) the java unit tests reside in the 'src/test/java' directory.
2, To make it more intuitive and easier to orienteering we usually use the same package for the test as the class under testing (in this case: 'main.services') 
3a, Uppercase names (like USER_SERVICE or RIGHT_PASSWORD) in Java indicates that the variable meant to be a constant ie. a variable that does not change value (and state). To reach this with Java String I suggest to use 'final', like: 
private static final String RIGHT_PASSWORD = "testPassword#123";
3b, USER_SERVICE is not a constant so I suggest _not_ to use uppercase in this case, rather camel case like userService. 
Otherwise, nice tests: one assert / function, clear, easy to understand.

**Pull request merged**